pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,353 @@
+	username := nsTmplSet.GetName()
+	nsName := ToNamespaceName(username, tcNamespace.Type)
+
+	tmplContent, err := templates.GetTemplateContent(fmt.Sprintf("%s-%s", nsTmplSet.Spec.TierName, tcNamespace.Type))
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	tmplProcessor := template.NewProcessor(r.client, r.scheme)
+	objs, err := tmplProcessor.Process(tmplContent, params)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+	objs = template.Filter(objs, template.RetainAllButNamespaces)
+	err = tmplProcessor.Apply(objs)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	namespace := &corev1.Namespace{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: nsName}, namespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to get namespace '%s'", nsName)
+	}
+	if namespace.Labels == nil {
+		namespace.Labels = make(map[string]string)
+	}
+	namespace.Labels["revision"] = tcNamespace.Revision
+	if err := r.client.Update(context.TODO(), namespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to update namespace '%s'", nsName)
+	}
+
+	log.Info("namespace provisioned", "namespace", tcNamespace)
+	if err := r.setStatusProvisioning(nsTmplSet); err != nil {
+		return err
+	}
+
+	// TODO add validation for other objects
+	return nil
+}
+
+func nextMissingNamespace(tcNamespaces []toolchainv1alpha1.Namespace, namespaces []corev1.Namespace, username string) (*toolchainv1alpha1.Namespace, *corev1.Namespace, bool) {
+	for _, tcNamespace := range tcNamespaces {
+		nsName := ToNamespaceName(username, tcNamespace.Type)
+		namespace, found := findNamespace(namespaces, nsName)
+		if found {
+			if namespace.Status.Phase == corev1.NamespaceActive && namespace.GetLabels()["revision"] != tcNamespace.Revision {
+				return &tcNamespace, &namespace, true
+			}
+		} else {
+			return &tcNamespace, nil, true
+		}
+	}
+	return nil, nil, false
+}
+
+func findNamespace(namespaces []corev1.Namespace, namespaceName string) (corev1.Namespace, bool) {
+	for _, ns := range namespaces {
+		if ns.GetName() == namespaceName {
+			return ns, true
+		}
+	}
+	return corev1.Namespace{}, false
+}
+
+// ToNamespaceName returns NamespaceName formed using given userName and nsType
+func ToNamespaceName(userName, nsType string) string {
+	return fmt.Sprintf("%s-%s", userName, nsType)
```

Comment:
Yes, Namespace name is owned by template.

But we want namespace_name before the template is loaded.  For ex here [code_link](https://github.com/codeready-toolchain/member-operator/blob/fa74b1fa6af80d3fafdb2259a767d5b2a7e5e897/pkg/controller/nstemplateset/nstemplateset_controller.go#L159).  Also, name got using `ToNamespaceName()` only used in logging purpose only.

Reply:
> But we want namespace_name before the template is loaded. For ex here code_link. 

Why do we need the name before the template is loaded? I'm probably missing something. I believe that you can load the template before that.

> Also, name got using ToNamespaceName() only used in logging purpose only.

that is not true - it's also used here: https://github.com/codeready-toolchain/member-operator/pull/53/files#diff-cffea82f7fdc7427b8c9f0d0fdd55aedR252

Reply:
why do you need to export this func? Also, the name of this func is confusing, since there's a k8s type called `NamespacedName` (with a `d`), which is totally different (in `k8s.io/apimachinery/api/types`)

Reply:
It will be needed for e2e test.

See this in member-operator: https://github.com/codeready-toolchain/member-operator/blob/master/pkg/controller/useraccount/useraccount_controller.go#L419
But re-implemented here: https://github.com/codeready-toolchain/toolchain-e2e/blob/master/e2e/e2e_test.go#L226

Reply:
Ideally, the e2e tests shouldn't have dependencies to the operator repos - only to api and common repos.

Anyway, I would love to avoid such a function and hard-coded generation of the namespace names. I do remember that there was a proposal to change the `username_dev` namespace to `username` and then, you would need to have the if-else logic same as we have in tenant-service but:
* to be honest, having such conditions for namespace names was a pain in the ....
* there will be the default namespace applied for namespaces created by advanced user and in that case, you won't be able to use this pattern

that said, I would rather use dynamic way of taking the namespace names and if possible use it from templates and do not hardcode it in the code.

Reply:
Yeah. I don't like the approach of hardcoding the namespace names in the code either. Currently it's used for logging/status updated but also for finding the next NS to provision as Matous mentioned.
But looking for the next namespace to provision using the namespace name from the template can cause issues too. What if we rename the namespace in a new version of the template? Let's say it was `{USERNAME}-ide` and in the new template it's `{USERNAME}-code`.
Maybe we should not use namespace names at all? Let the template processor to create NSs with whatever names are specified in the template. And for finding the namespace to provision in the controller we can use Template/NS type annotation (or label?): "code", "dev", "stage", etc. It would be probably more flexible approach.
For logging/status updates we could load the NS names from the processed template.

Reply:
@alexeykazakov thanks for solution.  I am going to add `type` label which takes values `code`, `dev` etc.  @MatousJobanek thanks to bring this up.

Reply:
Done https://github.com/codeready-toolchain/member-operator/pull/53/commits/68b4a33f18e24002da468750ddcfeb9a8d19cc04

When removing usage from test, I am getting some error to list_namespace with matched_labels (label with owner and type).  For now, I have hard_coded the namespace_name logic but revisit to remove it.  I spent good time but not able to find why it's not working.  I have commented the code which giving error and also added TODO explaining error details.  Any help would be appreciated, [code_link](https://github.com/codeready-toolchain/member-operator/pull/53/commits/68b4a33f18e24002da468750ddcfeb9a8d19cc04#diff-9b35dfc60ea38cb3e546642f4f9cedb4R263-R275)

Reply:
@nurali-techie maybe some fake client limitation? I wonder if it can't convert the object into a namespace when listing...

Reply:
@xcoulon knows better here, he helped me on this and he has more idea.

Reply:
actually, the fake client does not work well when using its `List()` func: if the objects passed during the calls to `Create()` were of the `unstructured.Unstructured` type, then the fake client's underlying object tracker is not able to convert these objects back to `corev1.Namespaces`, etc. This may result in an error. The current workaround is to explicitly convert the objects to create to `corev1.Namespaces` in the business code before calling the `client.Create()` func. I don't think it would do arm to the production code, and it may bypass with the fake client's limitations.